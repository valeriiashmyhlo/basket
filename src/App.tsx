import React from "react";
import "./App.css";
import { Col, Container, Row } from "react-bootstrap";
import { ItemList } from "./components/ItemList";
import { Cart } from "./components/Cart";
import { Item, CartItem } from "./types";
import { NForPrice, DefaultRule, NForM } from "./priceRules";

const GOODS: { [index: string]: Item } = {
  1: {
    id: "1",
    name: "Face Mask",
    priceRule: new NForPrice(2, 2.5, 4.0),
  },
  2: {
    id: "2",
    name: "Toilet Paper",
    priceRule: new NForM(6, 5, 0.65),
  },
  3: {
    id: "3",
    name: "Hand Sanitizer",
    priceRule: new DefaultRule(19.99),
  },
};

const App = () => {
  const [cartList, setCartList] = React.useState<{ [index: string]: CartItem }>(
    {}
  );

  const onAddToCart = (id: string) => {
    setCartList((prevState) => {
      const cartItem =
        id in prevState
          ? { ...prevState[id], quantity: prevState[id].quantity + 1 }
          : { item: GOODS[id], quantity: 1 };

      return {
        ...prevState,
        [id]: cartItem,
      };
    });
  };

  return (
    <div className="App">
      <Container>
        <Row>
          <Col>
            <ItemList
              itemList={Object.values(GOODS)}
              onAddToCart={onAddToCart}
            />
          </Col>
          <Col xs lg="3">
            <Cart items={Object.values(cartList)} />
          </Col>
        </Row>
      </Container>
    </div>
  );
};

export default App;
