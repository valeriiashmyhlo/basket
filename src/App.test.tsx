import React from "react";
import { fireEvent, render } from "@testing-library/react";
import App from "./App";

test("renders App component", () => {
  const { container } = render(<App />);
  expect(container.firstChild).toMatchSnapshot();
});

test("elements added to cart onClick add to cart", () => {
  const { container, getAllByText } = render(<App />);
  expect(container.firstChild).toMatchSnapshot();
  fireEvent.click(getAllByText("Add to cart")[0]);
  expect(container.firstChild).toMatchSnapshot();
});
