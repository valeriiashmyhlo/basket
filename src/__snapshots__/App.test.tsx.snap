// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elements added to cart onClick add to cart 1`] = `
<div
  class="App"
>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="itemList"
        >
          <div
            class="card"
            style="width: 175px; margin: 10px;"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                Face Mask
              </div>
              <p
                class="card-text"
              >
                2.5
              </p>
              <button
                class="btn btn-primary"
                type="button"
              >
                Add to cart
              </button>
            </div>
          </div>
          <div
            class="card"
            style="width: 175px; margin: 10px;"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                Toilet Paper
              </div>
              <p
                class="card-text"
              >
                0.65
              </p>
              <button
                class="btn btn-primary"
                type="button"
              >
                Add to cart
              </button>
            </div>
          </div>
          <div
            class="card"
            style="width: 175px; margin: 10px;"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                Hand Sanitizer
              </div>
              <p
                class="card-text"
              >
                19.99
              </p>
              <button
                class="btn btn-primary"
                type="button"
              >
                Add to cart
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-lg-3 col"
      >
        <div>
          <div>
            Your cart is empty
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`elements added to cart onClick add to cart 2`] = `
<div
  class="App"
>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="itemList"
        >
          <div
            class="card"
            style="width: 175px; margin: 10px;"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                Face Mask
              </div>
              <p
                class="card-text"
              >
                2.5
              </p>
              <button
                class="btn btn-primary"
                type="button"
              >
                Add to cart
              </button>
            </div>
          </div>
          <div
            class="card"
            style="width: 175px; margin: 10px;"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                Toilet Paper
              </div>
              <p
                class="card-text"
              >
                0.65
              </p>
              <button
                class="btn btn-primary"
                type="button"
              >
                Add to cart
              </button>
            </div>
          </div>
          <div
            class="card"
            style="width: 175px; margin: 10px;"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                Hand Sanitizer
              </div>
              <p
                class="card-text"
              >
                19.99
              </p>
              <button
                class="btn btn-primary"
                type="button"
              >
                Add to cart
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-lg-3 col"
      >
        <div>
          <div
            class="cartItem my-2"
          >
            <span>
              Face Mask
            </span>
            <span>
              x 
              1
            </span>
            <span>
              Â£
              2.5
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders App component 1`] = `
<div
  class="App"
>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="itemList"
        >
          <div
            class="card"
            style="width: 175px; margin: 10px;"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                Face Mask
              </div>
              <p
                class="card-text"
              >
                2.5
              </p>
              <button
                class="btn btn-primary"
                type="button"
              >
                Add to cart
              </button>
            </div>
          </div>
          <div
            class="card"
            style="width: 175px; margin: 10px;"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                Toilet Paper
              </div>
              <p
                class="card-text"
              >
                0.65
              </p>
              <button
                class="btn btn-primary"
                type="button"
              >
                Add to cart
              </button>
            </div>
          </div>
          <div
            class="card"
            style="width: 175px; margin: 10px;"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                Hand Sanitizer
              </div>
              <p
                class="card-text"
              >
                19.99
              </p>
              <button
                class="btn btn-primary"
                type="button"
              >
                Add to cart
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-lg-3 col"
      >
        <div>
          <div>
            Your cart is empty
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
